cpp/velox/shuffle/VeloxSplitter.cc

```diff
@@ -0,0 +1,1351 @@
+  uint32_t size_per_row = 0;
+  auto num_rows = rv.size();
+  for (size_t i = fixed_width_column_count_; i < simple_column_indices_.size(); ++i) {
+    auto index = i - fixed_width_column_count_;
+    if (binary_array_empirical_size_[index] == 0) {
+      auto column = rv.childAt(simple_column_indices_[i]);
+      auto string_view_column = column->asFlatVector<velox::StringView>();
+      assert(string_view_column);
+
+      // accumulate length
+      uint64_t length = 0;
+      auto string_views = string_view_column->rawValues<velox::StringView>();
+      for (size_t row = 0; row != num_rows; ++row) {
+        length += string_views[row].size();
+      }
+
+      binary_array_empirical_size_[index] = length % num_rows == 0 ? length / num_rows : length / num_rows + 1;
+    }
+  }
+
+  size_per_row = std::accumulate(binary_array_empirical_size_.begin(), binary_array_empirical_size_.end(), 0);
+
+  for (size_t col = 0; col < simple_column_indices_.size(); ++col) {
+    auto col_idx = simple_column_indices_[col];
+    size_per_row += arrow::bit_width(arrow_column_types_[col_idx]->id()) >> 3;
+  }
+
+  uint64_t prealloc_row_cnt = options_.offheap_per_task > 0 && size_per_row > 0
+      ? options_.offheap_per_task / size_per_row / num_partitions_ >> 2
+      : options_.buffer_size;
+  prealloc_row_cnt = std::min(prealloc_row_cnt, (uint64_t)options_.buffer_size);
+
+  return prealloc_row_cnt;
+}
+
+arrow::Status VeloxSplitter::AllocateBufferFromPool(std::shared_ptr<arrow::Buffer>& buffer, uint32_t size) {
+  // if size is already larger than buffer pool size, allocate it directly
+  // make size 64byte aligned
+  auto reminder = size & 0x3f;
+  size += (64 - reminder) & ((reminder == 0) - 1);
+  if (size > SPLIT_BUFFER_SIZE) {
+    ARROW_ASSIGN_OR_RAISE(buffer, arrow::AllocateResizableBuffer(size, options_.memory_pool.get()));
+    return arrow::Status::OK();
+  } else if (combine_buffer_->capacity() - combine_buffer_->size() < size) {
+    // memory pool is not enough
+    ARROW_ASSIGN_OR_RAISE(
+        combine_buffer_, arrow::AllocateResizableBuffer(SPLIT_BUFFER_SIZE, options_.memory_pool.get()));
+    combine_buffer_->Resize(0, /*shrink_to_fit = */ false);
+  }
+  buffer = arrow::SliceMutableBuffer(combine_buffer_, combine_buffer_->size(), size);
+  combine_buffer_->Resize(combine_buffer_->size() + size, /*shrink_to_fit = */ false);
+  return arrow::Status::OK();
+}
+
+arrow::Status VeloxSplitter::AllocatePartitionBuffers(uint32_t partition_id, uint32_t new_size) {
+  // try to allocate new
+  auto num_fields = schema_->num_fields();
+  assert(num_fields == arrow_column_types_.size());
+
+  auto fixed_width_idx = 0;
+  auto binary_idx = 0;
+  auto list_idx = 0;
+
+  std::vector<std::shared_ptr<arrow::ArrayBuilder>> new_list_builders;
+
+  for (auto i = 0; i < num_fields; ++i) {
+    size_t sizeof_binary_offset = -1;
+    switch (arrow_column_types_[i]->id()) {
+      case arrow::BinaryType::type_id:
```

Comment:
I don't have much context here but my intuition is that the code can base on row vector's schema instead of Arrow schema. But current way is also OK to me as of now.

Reply:
> I don't have much context here but my intuition is that the code can base on row vector's schema instead of Arrow schema. But current way is also OK to me as of now.

I used RowVector schema first in fact,  but I found it's not a good method because too many changes are required.

Reply:
OK. So let's just keep the current way, it's not a big deal at this time.